### PR TITLE
feature: warn if modules is missing any mentioned modules 

### PR DESCRIPTION
**Pull request by Alizter**
We warn the user if modules_without_implementation, private_modules or
virtual_modules contains any modules not in the modules field.

Fixes https://github.com/ocaml/dune/issues/7026

This will be made into an error in 3.9.

### CODE REVIEW

**Review by rgrinberg — CHANGES_REQUESTED** [review]
This warning should be ignored for vendored directories.

**Comment by Alizter** [comment]
I've allowed the warning to be ignored when appearing inside a vendored directory and checked the vendored_dirs cases in each of the tests.

File: src/dune_rules/modules_field_evaluator.mli [hunk 1]
```diff
@@ -25,5 +25,6 @@ val eval :
   -> private_modules:Ordered_set_lang.t
   -> kind:kind
   -> src_dir:Path.Build.t
+  -> is_vendored:bool
```
Comment (rgrinberg) [inline src/dune_rules/modules_field_evaluator.mli:28]: We're already passing src_dir. Can we check if it's vendored inside eval?

File: src/dune_rules/modules_field_evaluator.ml [hunk 2]
```diff
@@ -213,6 +227,21 @@ let check_invalid_module_listing ~stanza_loc ~modules_without_implementation
       (unimplemented_virt_modules |> Module_name.Path.Set.to_list
       |> List.map ~f:(fun name -> (stanza_loc, name)))
       [ Pp.text "You must provide an implementation for all of these modules." ];
+    (* Checking that (modules) incldues all declared modules *)
```
Comment (rgrinberg) [inline src/dune_rules/modules_field_evaluator.ml:230]: `includes.

File: src/dune_rules/modules_field_evaluator.ml [hunk 3]
```diff
@@ -213,6 +227,21 @@ let check_invalid_module_listing ~stanza_loc ~modules_without_implementation
...
+    let print_undelared_modules field mods =
+      (* TODO: this is a warning for now, change to an error in 3.9. *)
+      (* If we are in a vendored stanza we do nothing. *)
+      if is_vendored then ()
```
Comment (rgrinberg) [inline src/dune_rules/modules_field_evaluator.ml:234]: `if not is_vendored ...`

File: src/dune_rules/modules_field_evaluator.mli (hunk 1 above)
Comment (rgrinberg) [reply src/dune_rules/modules_field_evaluator.mli:28]: I don't see an issue with that.

**Review by rgrinberg — APPROVED** [review]

**Comment by Alizter** [comment]
@rgrinberg merging this when the CI is green.

**Pull request merged**